pkg/server/server.go

```diff
@@ -738,14 +740,27 @@ func (s *ProxyServer) serveRecvBackend(backend Backend, stream agent.AgentServic
 			} else {
 				dialErr := false
 				if resp.Error != "" {
-					klog.ErrorS(errors.New(resp.Error), "DIAL_RSP contains failure", "dialID", resp.Random, "agentID", agentID, "connectionID", resp.ConnectID)
+					// Dial response with error should not contain a valid ConnID.
+					klog.ErrorS(errors.New(resp.Error), "DIAL_RSP contains failure", "dialID", resp.Random, "agentID", agentID)
 					dialErr = true
 				}
 				err := frontend.send(pkt)
 				s.PendingDial.Remove(resp.Random)
 				if err != nil {
 					klog.ErrorS(err, "DIAL_RSP send to frontend stream failure",
 						"dialID", resp.Random, "serverID", s.serverID, "agentID", agentID, "connectionID", resp.ConnectID)
+					// If we never finish setting up the tunnel for ConnectID, then the connection is dead.
+					// Currently, the agent will no resend DIAL_RSP, so connection is dead.
+					// We already attempted to tell the frontend that. We should ensure we tell the backend.
+					pkt := &client.Packet{
+						Type: client.PacketType_CLOSE_REQ,
+						Payload: &client.Packet_CloseRequest{
+							CloseRequest: &client.CloseRequest{
+								ConnectID: resp.ConnectID,
+							},
+						},
+					}
+					_ = stream.Send(pkt)
 					dialErr = true
 				}
 				// Avoid adding the frontend if there was an error dialing the destination
```

Comment:
Doesn't the break below break out of the whole Receive loop, and drop all connections to the agent? That doesn't seem right....

Reply:
I think it should be a continue? That seems like a pretty major bug...

Reply:
Silly me... I forgot that `break` exits a switch statement, not just loops :facepalm: